### PR TITLE
Torch deepseek v2

### PR DESCRIPTION
**Pull request by grimoire**
- MLA implementation hinted by https://kexue.fm/archives/10091 . kv share same cache blocks.
- q,k will be loaded by two blocks(triton require block pow2)
- `q_a_proj`, `kv_a_proj_with_mqa` in attention layer, `gate` in moe layer are not distributed so less nccl op are required with the cost of memorys.
- Each GPU takes a response for 20 experts (8 A100).
- `block_size=32` would have better performance.
- large amount of runtime memories are required. big `cache_max_entry_count` and `max_prefill_token_num` might leads to oom.

result of deepseek-v2-lite (WIP)

```bash
python3 benchmark/profile_throughput.py \
    ShareGPT_V3_unfiltered_cleaned_split.json \
    DeepSeek-V2-Lite-Chat \
    --backend pytorch \
    --cache-max-entry-count 0.8 \
    --num-prompts 3000 \
    --cache-block-seq-len 32 \
    --concurrency 256
```

```
--------------------------------------------------
concurrency: 256
elapsed_time: 311.155s

first token latency(s)(min, max, ave): 2.337, 19.513, 4.276
per-token latency(s) percentile(50, 75, 95, 99): [0.056, 0.112, 0.311, 0.408]

number of prompt tokens: 721790
number of completion tokens: 668262
token throughput (completion token): 2147.679 token/s
token throughput (prompt + completion token): 4467.387 token/s
RPS (request per second): 9.641 req/s
RPM (request per minute): 578.489 req/min
--------------------------------------------------
```

requirements:

- [x] #1520 
- [x] #1603

### CODE REVIEW

**Comment by grimoire** [comment]
Since k has size 512+64, the block size of MHA should be 1024 (triton require pow2 block size). Paged attention kernel take too much time.

File: lmdeploy/pytorch/configurations/deepseek_v2.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    def condition(cls, hf_config):
+        """config."""
+        return hf_config.model_type == 'deepseek_v2'
+
+    @classmethod
+    def build(cls, hf_config):
+        """build."""
+        head_dim = (hf_config.kv_lora_rank + hf_config.qk_rope_head_dim)
+        k_head_dim = head_dim
+        v_head_dim = 0
```
Comment (RunningLeon) [inline lmdeploy/pytorch/configurations/deepseek_v2.py:19]: why `v_head_dim` is `0`?

Comment (grimoire) [reply lmdeploy/pytorch/configurations/deepseek_v2.py:19]: https://kexue.fm/archives/10091

**Comment by zhyncs** [comment]
Hi @lvhan028 After this PR is ready and merged, will LMDeploy release a new release? Thanks. @grimoire

**Comment by zhyncs** [comment]
> result of deepseek-v2-lite

Hi @grimoire Are the current performance benchmark results as expected, and how much of a leading advantage is there compared to vLLM? Thanks.

https://github.com/deepseek-ai/DeepSeek-V2?tab=readme-ov-file#inference-with-vllm-recommended

**Comment by grimoire** [comment]
@zhyncs the latest profile result (256 concurrency, 3000 prompt, block_size=32 and `--cache-max-entry-count` has been adjusted to prevent OOM):
- deepseek v2: 3.627 req/s
- deepseek v2 lite: 10.404 req/s

Apart from the fact that the default value cannot be used for block_size, the rest is relatively acceptable.

We have not performed benchmarks on vLLM yet, 8 A100 are not always available (T T).

**Comment by RunningLeon** [comment]
@grimoire ut `TestMBGMV.test_mbgmv` failed. may need to fix.

**Comment by zhyncs** [comment]
Hi @grimoire, I used your commit to run the workflow at https://github.com/zhyncs/lmdeploy/actions/runs/9584655537 and obtained the whl https://github.com/zhyncs/dl/releases/tag/0620. And I encountered an error https://github.com/triton-lang/triton/issues/4172. Do you have any ideas? Thanks!

**Comment by grimoire** [comment]
triton has a prepackaged ptxas, which might be different with your cuda driver version. You can set your own ptxas (`/path/to/cuda/bin/ptxas`) with environment `TRITON_PTXAS_PATH`.

**Comment by zhyncs** [comment]
```bash
export TRITON_PTXAS_PATH=/usr/local/cuda/bin/ptxas
```
It works for me. Thanks and cheers. @grimoire

**Comment by zhyncs** [comment]
> * deepseek v2 lite: 10.404 req/s

Hi @grimoire May I ask if this uses a single A100 card or 8 cards? Thanks.

**Comment by zhyncs** [comment]
LMDeploy
https://github.com/zhyncs/dl/releases/tag/0620
https://github.com/zhyncs/lmdeploy/actions/runs/9584655537
https://github.com/grimoire/lmdeploy/commit/995e0ed972a024445cfb9ae44b292e4907c4a798

**single A100**
```bash
# server
python3 -m lmdeploy serve api_server DeepSeek-V2-Lite --backend pytorch --cache-block-seq-len 32

# client
# https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py
python3 benchmark_serving.py --backend lmdeploy --host 127.0.0.1 --port 23333 --dataset /workdir/ShareGPT_V3_unfiltered_cleaned_split.json --model DeepSeek-V2-Lite --tokenizer DeepSeek-V2-Lite --num-prompts 1000 --request-rate 128
```

result
```bash
# ignore_eos false
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  154.05
Total input tokens:                      236142
Total generated tokens:                  148682
Request throughput (req/s):              6.49
Input token throughput (tok/s):          1532.88
Output token throughput (tok/s):         965.14
---------------Time to First Token----------------
Mean TTFT (ms):                          56583.14
Median TTFT (ms):                        55727.01
P99 TTFT (ms):                           113475.30
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          116.80
Median TPOT (ms):                        90.45
P99 TPOT (ms):                           475.46
---------------Inter-token Latency----------------
Mean ITL (ms):                           77.64
Median ITL (ms):                         58.83
P99 ITL (ms):                            430.49
==================================================

# ignore_eos true
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  181.48
Total input tokens:                      236142
Total generated tokens:                  215605
Request throughput (req/s):              5.51
Input token throughput (tok/s):          1301.17
Output token throughput (tok/s):         1188.01
---------------Time to First Token----------------
Mean TTFT (ms):                          65216.61
Median TTFT (ms):                        65241.68
P99 TTFT (ms):                           135946.04
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          96.65
Median TPOT (ms):                        80.46
P99 TPOT (ms):                           267.16
---------------Inter-token Latency----------------
Mean ITL (ms):                           72.56
Median ITL (ms):                         60.45
P99 ITL (ms):                            372.03
==================================================
```

**Comment by grimoire** [comment]
> Hi @grimoire May I ask if this uses a single A100 card or 8 cards? Thanks.

It is profiled with single A100. The bottleneck of lite model is on the host side, TP would make it worse.

**Comment by zhyncs** [comment]
> * `block_size=32` would have better performance.

May we set the `cache-block-seq-len` to 32 by default when running DeepSeek V2 for inference? From the benchmark results, there is a significant performance gap.

```bash
# python3 benchmark_serving.py --backend lmdeploy --host 127.0.0.1 --port 23333 --dataset /workdir/ShareGPT_V3_unfiltered_cleaned_split.json --model DeepSeek-V2-Lite --tokenizer DeepSeek-V2-Lite --num-prompts 1000 --request-rate 128
# cache-block-seq-len 32, ignore_eos true
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  181.48
Total input tokens:                      236142
Total generated tokens:                  215605
Request throughput (req/s):              5.51
Input token throughput (tok/s):          1301.17
Output token throughput (tok/s):         1188.01
---------------Time to First Token----------------
Mean TTFT (ms):                          65216.61
Median TTFT (ms):                        65241.68
P99 TTFT (ms):                           135946.04
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          96.65
Median TPOT (ms):                        80.46
P99 TPOT (ms):                           267.16
---------------Inter-token Latency----------------
Mean ITL (ms):                           72.56
Median ITL (ms):                         60.45
P99 ITL (ms):                            372.03
==================================================


# cache-block-seq-len default, ignore_eos true
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  384.18
Total input tokens:                      236142
Total generated tokens:                  215594
Request throughput (req/s):              2.60
Input token throughput (tok/s):          614.67
Output token throughput (tok/s):         561.18
---------------Time to First Token----------------
Mean TTFT (ms):                          155387.50
Median TTFT (ms):                        153036.64
P99 TTFT (ms):                           328194.63
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          196.80
Median TPOT (ms):                        181.42
P99 TPOT (ms):                           515.61
---------------Inter-token Latency----------------
Mean ITL (ms):                           163.64
Median ITL (ms):                         96.56
P99 ITL (ms):                            1542.84
==================================================
```

**Review by RunningLeon — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/engine/model_agent.py [hunk 2]
```diff
@@ -79,6 +79,21 @@ def __get_free_gpu_mem_size(cache_block_size: int):
                      f' {runtime_cache_size>>20} mb')
         return gpu_mem_physical_free * cache_config.cache_max_entry_count
 
+    def __adjust_block_size():
+        """adjust block_size."""
+        # TODO: support kernel with both large head dim and large block size.
+        if model_config.k_head_dim >= 512 and cache_config.block_size > 32:
+            cache_config.block_size = 32
```
Comment (zhyncs) [inline lmdeploy/pytorch/engine/model_agent.py:86]: Will this affect models other than DeepSeek v2?

Comment (grimoire) [reply lmdeploy/pytorch/engine/model_agent.py:86]: Yes, the mha kernel needs enough smem to cache the kv_cache block and query block. Any model with such a large head_dim should be limited.
Among all the models that Pytorch engine has supported, only deepseek v2 with MLA implementation meets the condition.

**Comment by grimoire** [comment]
@zhyncs temperature=0 is a invalid value
https://github.com/huggingface/transformers/blob/730a440734e1fb47c903c17e3231dac18e3e5fd6/src/transformers/generation/logits_process.py#L298

I set to 1 if temperature<=0 in pytorch engine.
https://github.com/InternLM/lmdeploy/blob/9e8cb3c4948c0160f00d5401d4519b192aee6581/lmdeploy/pytorch/messages.py#L75

File: lmdeploy/pytorch/check_env/__init__.py [hunk 3]
```diff
@@ -69,7 +73,7 @@ def check_env():
 
 
 MIN_TRANSFORMERS_VERSION = '4.33.0'
-MAX_TRANSFORMERS_VERSION = '4.38.2'
+MAX_TRANSFORMERS_VERSION = '4.41.2'
```
[the remark this answers is not in the document]
Comment (grimoire) [reply lmdeploy/pytorch/check_env/__init__.py:76]: We have removed version requirements for a long time.

[the remark this answers is not in the document]
Comment (lvhan028) [reply lmdeploy/pytorch/check_env/__init__.py:76]: oh, sorry. I forgot that.

**Comment by zhyncs** [comment]
@grimoire If the temperature is 0, it is not supported. How can I get a deterministic answer?

**Comment by zhyncs** [comment]
TurboMind supports [0,2] for temperature

**Comment by grimoire** [comment]
Just set topk=1 or given a small enough temperature
https://github.com/huggingface/transformers/blob/730a440734e1fb47c903c17e3231dac18e3e5fd6/src/transformers/generation/logits_process.py#L275-L279

Note that small temperature might still leads to different result if two value in logits are close.

File: lmdeploy/pytorch/tools/make_inputs.py [hunk 4]
```diff
@@ -157,11 +179,16 @@ def __forward_hook(module, args, kwargs, output):
             target_args = args
             target_kwargs = kwargs
             target_output = output
+            raise ExtractorFound()
```
Comment (lvhan028) [inline lmdeploy/pytorch/tools/make_inputs.py:182]: why?

Comment (grimoire) [reply lmdeploy/pytorch/tools/make_inputs.py:182]: This tool is used to extract input/output of a submodule (for debug), computation after the module is not necessary.

File: lmdeploy/pytorch/tools/make_inputs.py [hunk 5]
```diff
@@ -43,37 +43,62 @@ def make_step_context(
...
+    world_size: int = 1,
 ):
     """make step context."""
+    device = input_ids.device
     from torch.nn.utils.rnn import pad_sequence
 
+    from lmdeploy.pytorch.engine.cache_engine import CacheEngine
     from lmdeploy.pytorch.engine.model_agent import StepContext
 
+    if model_config is None:
```
Comment (lvhan028) [inline lmdeploy/pytorch/tools/make_inputs.py:58]: Should we assert `model_config`

**Review by lvhan028 — APPROVED** [review]

File: lmdeploy/pytorch/models/deepseek_v2.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                q = self.q_b_proj(
+                    self.q_a_layernorm(self.q_a_proj(hidden_states)))
+            q = q.view(q_len, num_heads, self.q_head_dim)
+            # q_pe: (q_len, num_heads, qk_rope_head_dim)
+            q_nope, q_pe = torch.split(
+                q, [self.qk_nope_head_dim, self.qk_rope_head_dim], dim=-1)
+            # q_nope: (q_len, num_heads, kv_lora_rank)
+            q_nope_out = query_states[..., :nope_size]
+            torch.bmm(q_nope.transpose(0, 1),
+                      self.w_kc,
```
Comment (ispobock) [inline lmdeploy/pytorch/models/deepseek_v2.py:129]: Hi @grimoire, have you considered how to absorb the `w_kc` into q_proj's weights offline instead of computing it in the online forward step?

Comment (grimoire) [reply lmdeploy/pytorch/models/deepseek_v2.py:129]: Merge weight of bmm required too many memory.

Comment (grimoire) [reply lmdeploy/pytorch/models/deepseek_v2.py:129]: Like this?

```python
import torch

def naive(a, w0, w1):
    num_heads = w1.size(0)
    mem_size = w0.numel() * w0.element_size() + w1.numel() * w0.element_size()
    print(f'naive memory size: {mem_size}')
    out = a@w0
    out = out.unflatten(1, (num_heads, -1)) # (seq, num_heads, w0_size)
    out = torch.bmm(out.transpose(0, 1), w1)
    return out.transpose(0, 1)


def fused(a, w0, w1):
    num_heads = w1.size(0)
    w0 = w0.unflatten(1, (num_heads, -1))
    w2 = torch.bmm(w0.transpose(0, 1), w1)
    w2 = w2.transpose(0, 1).flatten(1, 2)
    mem_size = w2.numel() * w2.element_size()
    print(f'fused memory size: {mem_size}')
    out = a@w2
    out = out.unflatten(-1, (num_heads, -1))
    return out


def main():
    in_size = 8
    num_heads = 16
    w0_size = 128
    w1_size = 512
    a = torch.rand(3, in_size)
    w0 = torch.rand(in_size, num_heads * w0_size)
    w1 = torch.rand(num_heads, w0_size, w1_size)

    naive_out = naive(a, w0, w1)
    fused_out = fused(a, w0, w1)

    torch.testing.assert_close(naive_out, fused_out)


if __name__ == '__main__':
    main()

```
